src/gp/kernel.jl

```diff
@@ -201,6 +201,25 @@ pw(k::RQ, x::AV) = _rq.(pw(SqEuclidean(), x), k.α)
 
 
 
+"""
+    Cosine <: Kernel
+
+Cosine Kernel with period parameter `p`.
+"""
+struct Cosine{Tp<:Real} <: Kernel
+    p::Tp
+end
+
+# Binary methods.
+ew(k::Cosine, x::AV, x′::AV) = cos.(pi.*ew(Euclidean(), x, x′) ./k.p)
+pw(k::Cosine, x::AV, x′::AV) = cos.(pi.*pw(Euclidean(), x, x′) ./k.p)
+
+# Unary methods.
+ew(k::Cosine, x::AV) = ones(eltype(x), length(x))
+pw(k::Cosine, x::AV) = pw(k, x, x)
```

Comment:
I've also just realised that we would be better off replacing this with something like
```julia
pw(k::Cosine, x::AV) = cos.(pi .* pw(Euclidean(), x) ./ k.p)
```
because `pw(Euclidean(), x)` is faster and more accurate than `pw(Euclidean(), x, x)` i.e. it's able to specialise on the fact that both arguments are the same.

Reply:
I will make these changes

Reply:
Great. Would make sense to make sure the tests are passing on your machine before pushing to make sure that we've caught everything :)

Reply:
Thank you I have made the changes 
running [test/runtests.jl](https://github.com/willtebbutt/Stheno.jl/blob/master/test/runtests.jl) now and will push it once all the tests have passed

Reply:
Do feel free to push if tests are failing and you're not sure why :)

Reply:
Okay I have pushed the changes but I'm getting this error while running the tests
```julia
p=1.0: Error During Test at /home/ritesh/Desktop/github/Stheno.jl/test/runtests.jl:10
  Got exception outside of a @test
  MethodError: no method matching differentiable_kernel_tests(::Cosine{Float64}, ::Array{Float64,1}, ::Array{Float64,2}, ::Array{Float64,2}, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1})
  Closest candidates are:
    differentiable_kernel_tests(!Matched::Kernel, ::AbstractArray{#s196,1} where #s196<:Real, ::AbstractArray{#s195,2} where #s195<:Real, ::AbstractArray{#s194,2} where #s194<:Real, ::AbstractArray{T,1} where T, ::AbstractArray{T,1} where T, ::AbstractArray{T,1} where T; rtol, atol) at /home/ritesh/Desktop/github/Stheno.jl/test/test_util.jl:287
    differentiable_kernel_tests(!Matched::AbstractRNG, !Matched::Kernel, !Matched::AbstractArray{T,1} where T, !Matched::AbstractArray{T,1} where T, ::AbstractArray{T,1} where T; rtol, atol) at /home/ritesh/Desktop/github/Stheno.jl/test/test_util.jl:314
  Stacktrace:
   [1] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:92
   [2] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
   [3] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:92
   [4] top-level scope at /home/ritesh/.julia/packages/TimerOutputs/7Id5J/src/TimerOutput.jl:214
   [5] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:91
   [6] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
   [7] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:91
   [8] top-level scope at /home/ritesh/.julia/packages/TimerOutputs/7Id5J/src/TimerOutput.jl:214
   [9] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:90
   [10] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
   [11] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:9
   [12] top-level scope at /home/ritesh/.julia/packages/TimerOutputs/7Id5J/src/TimerOutput.jl:214
   [13] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:8
   [14] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
   [15] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/gp/kernel.jl:8
   [16] top-level scope at /home/ritesh/.julia/packages/TimerOutputs/7Id5J/src/TimerOutput.jl:214
   [17] include at ./boot.jl:328 [inlined]
   [18] include_relative(::Module, ::String) at ./loading.jl:1105
   [19] include(::Module, ::String) at ./Base.jl:31
   [20] include(::String) at ./client.jl:424
   [21] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/runtests.jl:33
   [22] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
   [23] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/runtests.jl:32
   [24] top-level scope at /home/ritesh/.julia/packages/TimerOutputs/7Id5J/src/TimerOutput.jl:214
   [25] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/runtests.jl:31
   [26] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
   [27] top-level scope at /home/ritesh/Desktop/github/Stheno.jl/test/runtests.jl:17
   [28] include at ./boot.jl:328 [inlined]
   [29] include_relative(::Module, ::String) at ./loading.jl:1105
   [30] include(::Module, ::String) at ./Base.jl:31
   [31] exec_options(::Base.JLOptions) at ./client.jl:287
   [32] _start() at ./client.jl:460
```
I'm not sure why Cosine isn't being recognized as a kernel

Reply:
I agree -- we might as well not have a `p` parameter in the `Cosine` kernel as the `stretch` transformation should just handle it.

Not sure what's going on in your above example, it doesn't seem to be having the same issue in CI. How are you running the tests?

Reply:
In the test folder of my local repository I'm just running
```julia
$ julia runtests.jl
```
on my terminal. Am I doing the correct thing?

Reply:
You might not have the right environment activated. It's generally best to navigate to the top level of the Stheno dev folder, start julia, and run
```julia
]activate .
]test
```

Reply:
Typing `]` activates the package manager

Reply:
Thank you very much and sorry for my late response, I was busy with another project. I have tested the kernel for various values of ```p``` and encountered two types of test failures
* The test ```     @test fd_isapprox(adj_ad, adj_fd, rtol, atol)``` in [line 90](https://github.com/willtebbutt/Stheno.jl/blob/61236affc381d5cba16fab67b9a0d93226e8995d/test/test_util.jl#L90) of the test file fails for ```p=10.0```, But this can be resolved by increasing the ```_atol``` and ```_rtol``` values from ```1e-10``` to ```1e-8``` in [test_util.jl](https://github.com/willtebbutt/Stheno.jl/blob/61236affc381d5cba16fab67b9a0d93226e8995d/test/test_util.jl#L8) i.e by loosening the error bound 
* The test ``` @test all(eigvals(Matrix(pw(k, x0))) .> -atol)``` in [test_util.jl](https://github.com/willtebbutt/Stheno.jl/blob/61236affc381d5cba16fab67b9a0d93226e8995d/test/test_util.jl#L155) fails for some values of ```p``` mostly less than ```5.0``` (even for an increased error bound ```1e-8```). I'm not sure why the matrix isn't positive semi-definite for these cases. 
I printed the eigvals for ```p=2.0``` case and they are 
```[-1.2286515550839097, -0.4334855269256753, 1.2394032841422056, 2.43584181544076, 2.986891982426619]```. Here, the first two eigen values are -ve and of order 1 which isn't generally expected. I will look more into it tomorrow
*  What do you think of loosening the bound idea for the first point? Can we have ```alleigvals > -1e-8``` instead of the current ```alleigvals > -1e-10``` ?

Reply:
Interesting.

- Numerical issues are to be expected with `fd_isapprox`, so I would just set different values for `rtol` and `atol` for that test.
- In terms of the eigenvalues, I think the issue is that the cosine kernel is only defined on the real line, not in higher dimension, so we should expect it to have all positive eigenvalues in higher dimensions. As such, I would suggest restricting support to only `Real`s, and just testing them :)